### PR TITLE
Pylint updates, part two 

### PR DESCRIPTION
**Pull request by ana**
More pylint updates to prepare the update to version 2.12.2

References: https://github.com/avocado-framework/avocado/issues/5211

@beraldoleal as mentioned, I sent it again, I realized of a problem after pushing it and was in the middle of testing when you started to review it.

### CODE REVIEW

**Review by beraldoleal — COMMENTED** [review]
Hi @ana, this LGTM. Thanks.

My only concern here is with the argument rename (msg -> message).

Since this is an important method, I'm not sure if the docstring it was giving enough visibility to the users. And yes, this probably could be captured when introduced and not now. But, at this point, users are not being properly notified with a warning message when using it.

**Comment by ana** [comment]
> Hi @ana, this LGTM. Thanks.
> 
> My only concern here is with the argument rename (msg -> message).
> 
> Since this is an important method, I'm not sure if the docstring it was giving enough visibility to the users. And yes, this probably could be captured when introduced and not now. But, at this point, users are not being properly notified with a warning message when using it.

@beraldoleal Fine with the warning  message, but I don't understand how this warning message should be or where.

@richtja  you added the  pylint: disable=W0221 in `avocado/core/test.py` in commit
  as part of 8d5b4e1936beaa35c3c8743ad97bffe526ac6e4c
it mentions some LTS release, do you know if it's safe to remove it now?

**Comment by beraldoleal** [comment]
> @beraldoleal Fine with the warning message, but I don't understand how this warning message should be or where.

We could use a decorator to support both signatures for a while or since this is just adding a new one, adding the second argument and avoid the decorator. And warn with:
 
```python
warnings.warn(..., DeprecationWarning)
 ```

**Comment by richtja** [comment]
> @richtja you added the pylint: disable=W0221 in `avocado/core/test.py` in commit as part of [8d5b4e1](https://github.com/avocado-framework/avocado/commit/8d5b4e1936beaa35c3c8743ad97bffe526ac6e4c) it mentions some LTS release, do you know if it's safe to remove it now?

Hi @ana It looks like that this was meant to be removed in 69LTS, but we forget to do that. I found this [discussion](https://github.com/avocado-framework/avocado/pull/3228) about it.

The only reason for this warning was to get users time to prepare for this change, so IMO we can remove this now.

**Review by richtja — APPROVED** [review]
LGTM

**Comment by ana** [comment]
> > @beraldoleal Fine with the warning message, but I don't understand how this warning message should be or where.
> 
> We could use a decorator to support both signatures for a while or since this is just adding a new one, another option would be adding the second argument and avoid the decorator. And warn with:
> 
> ```python
> warnings.warn(..., DeprecationWarning)
> ```

Thanks for the suggestion @beraldoleal I just force-pushed an update.

File: avocado/core/test.py [hunk 1]
```diff
@@ -877,37 +879,50 @@ def _report(self):
...
         """
         Fails the currently running test.
 
         After calling this method a test will be terminated and have its status
         as FAIL.
 
-        :param message: an optional message that will be recorded in the logs
-        :type message: str
-        :warning message: This parameter will changed name to "msg" in the next
-                          LTS release because of lint W0221
```
Comment (beraldoleal) [inline avocado/core/test.py:890]: Since we are still accepting this argument, I would still keep this warning docstring.

**Review by beraldoleal — CHANGES_REQUESTED** [review]
Hi @ana this LGTM besides one small comment about the warning.

File: avocado/core/test.py (hunk 1 above)
Comment (ana) [reply avocado/core/test.py:890]: I think the current warning on runtime is enough but if you insist...

Comment (ana) [reply avocado/core/test.py:890]: @beraldoleal I've been thinking about this. People reading the documentation are likely to never used this class before, so they'll just use msg directly. And the warning won't be helpful.
For older users, the runtime warning I added already is the way to go.

Comment (beraldoleal) [reply avocado/core/test.py:890]: > People reading the documentation are likely to never used this class before. And the warning won't be helpful. For older users, the runtime warning I added already is the way to go.

I see the reasoning for removing this, however this might be a strong assumption.

Current users/contributors seeing the warning message at the terminal, might need to double check the method's documentation, to understand better before adapting their code. And here we see additional information, like the reason because it is being deprecated and when it will be removed.

IMO, the same is valid for new users/contributors, since they might be looking at mixed calls (msg vs message) on examples and third-party code, and it is trying to understand what is the right way to do, without the need to run a test. So, he/she might need to double check the documentation as well.

To be clear, this is not a "strong blocker" to me, and I'm not "insisting". I just would like to make sure that our documentation is in sync with our code.

Perhaps, I was not so clear the last time we spoke, and I apologize for that. But, maybe this will not be needed at all, as I tried to say before. So possibly, one way is just hold this commit until we double check the real need for this decorator vs "we are safe to remove it since we are not breaking our promise regarding the API compatibly.

Comment (ana) [reply avocado/core/test.py:890]: @beraldoleal I see.  When I added the decorator was with the expectation this PR could be merged fast.
 I think it's a very good solution  (*) because it handles both variable names and it doesn't break anything. It's a pity we didn't handle things this way from the beginning, but it's never late, we could keep this decorator and remove it after the next LTS.

(*) I expect almost a full day on this, so I'm clearly biased here!

File: avocado/core/nrunner.py [hunk 2]
```diff
@@ -100,11 +100,11 @@ class ConfigEncoder(json.JSONEncoder):
     JSON Encoder for config options.
     """
 
-    def default(self, config_option):  # pylint: disable=W0221
-        if isinstance(config_option, set):
-            return {'__encoded_set__': list(config_option)}
+    def default(self, o):
```
Comment (clebergnu) [inline avocado/core/nrunner.py:103]: Such a nice variable name in cPython... sigh. But your change looks good, thanks!

File: avocado/core/test.py [hunk 3]
```diff
@@ -877,37 +879,50 @@ def _report(self):
...
+        """ This decorator helps to deprecate parameter 'message' and
+        replace it with 'msg'.
+        """
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            message = kwargs.get('message', None)
+            if message is not None:
+                kwargs['msg'] = message
+                warnings.warn("The parameter 'message' has been deprecated. Please use instead 'msg'", DeprecationWarning)
```
Comment (clebergnu) [inline avocado/core/test.py:892]: Nitpick: consider breaking the huge string into two lines.

File: avocado/core/test.py [hunk 4]
```diff
@@ -877,37 +879,50 @@ def _report(self):
...
+            return func(*args, **kwargs)
+        return wrapper
+
+    @_deprecate_params_message
+    def fail(self, msg=None, **kwargs):
```
Comment (clebergnu) [inline avocado/core/test.py:897]: Not related to your changes, but the other decorated methods (`error`, `cancel`) are static and this one is not. We should make them consistent.

File: avocado/core/test.py (hunk 3 above)
Comment (clebergnu) [inline avocado/core/test.py:892]: One corner case not handled here seems to be if the user gives both `msg` and `message` parameters, with `msg` now being a possibility in either in positional or keyword form.  I think we can simplify the decorated methods, while still covering this corner case.  One positive aspect is that their signature will remain the same for the future, when the decorators are removed.

Suggestion, on top of your work:

```diff
diff --git a/avocado/core/test.py b/avocado/core/test.py
index 23a896448..20e8699ef 100644
--- a/avocado/core/test.py
+++ b/avocado/core/test.py
@@ -886,15 +886,32 @@ class Test(unittest.TestCase, TestData):
 
         @functools.wraps(func)
         def wrapper(*args, **kwargs):
-            message = kwargs.get('message', None)
-            if message is not None:
-                kwargs['msg'] = message
-                warnings.warn("The parameter 'message' has been deprecated. Please use instead 'msg'", DeprecationWarning)
-            return func(*args, **kwargs)
+            msg = None
+            message = None
+
+            if 'msg' in kwargs:
+                msg = kwargs.get('msg')
+            elif args:
+                msg = args[0]
+
+            if 'message' in kwargs:
+                message = kwargs.get('message')
+                if msg:
+                    warnings.warn("Please favor parameter 'msg' and do not use "
+                                  "it with the parameter 'message'",
+                                  DeprecationWarning)
+
+                else:
+                    warnings.warn("The parameter 'message' has been deprecated."
+                                  " Please use 'msg' instead", DeprecationWarning)
+
+            actual_message = msg or message
+            return func(actual_message)
         return wrapper
 
+    @staticmethod
     @_deprecate_params_message
-    def fail(self, msg=None, **kwargs):
+    def fail(msg=None):
         """
         Fails the currently running test.
 
@@ -908,7 +925,7 @@ class Test(unittest.TestCase, TestData):
 
     @staticmethod
     @_deprecate_params_message
-    def error(msg=None, **kwargs):
+    def error(msg=None):
         """
         Errors the currently running test.
 
@@ -922,7 +939,7 @@ class Test(unittest.TestCase, TestData):
 
     @staticmethod
     @_deprecate_params_message
-    def cancel(msg=None, **kwargs):
+    def cancel(msg=None):
         """
         Cancels the test.
 

```

I've tested it with this:

```python
from avocado import Test


class FailTest(Test):

    def test_empty(self):
        self.fail()

    def test_positional(self):
        self.fail('msg positional')

    def test_msg_keyword(self):
        self.fail(msg='msg keyword')

    def test_message_keyword(self):
        self.fail(message='msg keyword')

    def test_msg_message(self):
        self.fail('msg positional', message='message keyword')

```

**Review by clebergnu — APPROVED** [review]
@ana this LGTM. There are a few suggestions there, but it can be proposed/added on top of your version if your prefer.

File: avocado/core/test.py (hunk 3 above)
Comment (ana) [reply avocado/core/test.py:887]: Thanks @clebergnu , I like this one more!

**Comment by ana** [comment]
@beraldoleal I have updated this following the suggestions from @clebergnu , could you please tell me if it's good to get merged?
I only have one PR on top of this one and the pylint update will be finished =)

**Review by beraldoleal — APPROVED** [review]
